[jit][script] Handling for py2/py3 division differences

zou3519 opened the pull request:
- In Python 2, use of `/` (regardless of int/float/Tensor) causes a compiler error if
  `from __future__ import division` is not imported in the file.
- The / operator is universally set to do "true" division for integers
- Added a `prim::FloorDiv` operator because it is used in loop unrolling.

The error if users use '/' in python 2 without importing from __future__
occurs when building the JIT AST.

cc @apaszke @zdevito

apaszke commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -550,9 +550,35 @@ RegisterOperators reg2({
+    // - py2: (int, int) -> int
+    // - py3: (int, int) -> float
+    // XXX: These should probably not be the same op as aten::div
+#if PY_MAJOR_VERSION > 2
```
This file really shouldn't depend on the Python version, because it is part of the no-python build too (in which case we'll choose the Python 2 behavior!).

zdevito commented:
I think the fix should be:

* In Python 2, use of `/` should cause a compiler error if `from __future__ import division` is not imported in the file. 
* The `/` operator is universally set to do "true" division for integers, meaning it returns a float.
* (optional, for now) We add the `//` operator support to script.

zdevito commented on torch/csrc/jit/register_prim_ops.cpp:
```diff
@@ -550,9 +550,35 @@ RegisterOperators reg2({
+            push(stack, static_cast<double>(a) / static_cast<double>(b));
+            return 0;
+          };
+        }),
```
My mistake, div needs to return float, which would not happen if we used the macro.

zou3519 committed:
[jit][script] Handling for py2/py3 division differences

- In Python 2, use of `/` (regardless of int/float/Tensor) causes a compiler error if
  `from __future__ import division` is not imported in the file.
- The / operator is universally set to do "true" division for integers
- Added a `prim::FloorDiv` operator because it is used in loop unrolling.

The error if users use '/' in python 2 without importing from __future__
occurs when building the JIT AST.

zou3519 commented on test/expect/TestScript.test_loop_unroll_unused_counter.expect:
```diff
@@ -3,7 +3,7 @@ graph(%x : Dynamic) {
   %3 : int = prim::Constant[value=1]()
   %4 : int = prim::Constant[value=8]()
-  %5 : int = aten::div(%2, %4)
+  %5 : int = prim::FloorDiv(%2, %4)
```
@jamesr66a should this be exportable to ONNX? If so I can write a pass to change it to aten::div.

apaszke replied:
I don't think we need to support it, unless we also add support for `//`. It's only emitted by loop unrolling, which doesn't run before ONNX.

zou3519 commented:
This should be good for another review. I'm a little concerned about ONNX export though (I'm not sure if I should be) because of the following:

- ints get exported as scalar tensors to ONNX
- aten::div(int, int) gives a float now because `/` is truediv. However, aten::div(int scalar tensor, int scalar tensor) gives an int scalar tensor because of division semantics on tensors...

@jamesr66a should I be concerned?

zou3519 committed:
Forgot to add test files

zou3519 commented:
Had offline chat with @apaszke, I am going to fix the ONNX export

zdevito commented on torch/jit/frontend.py:
```diff
@@ -124,13 +124,40 @@ def build_stmts(ctx, stmts):
+
+        # Check if feature is a __future__ _Feature to make sure this isn't some
+        # other 'division' library. _Feature instances have the following attrs:
+        return hasattr(feature, 'getOptionalRelease') and hasattr(feature, 'getMandatoryRelease')
```
Does this work?
```
import __future__
return fn.__globals__.get('division') is __future__.division
```

zdevito reviewed (approved):
This looks good -- I have some minor comments.

zou3519 committed:
Address comments:
- s/prim::FloorDiv/aten::floordiv
- Fix onnx export for true division on integers

zou3519 committed:
Simply __future__ division check

apaszke commented on torch/csrc/jit/passes/prepare_division_for_onnx.h:
```diff
@@ -0,0 +1,17 @@
+// - aten::div(int, int) -> float is the python truediv operator. This doesn't
+//   exist in ONNX so we express it in terms of ONNX ops.
+//
+TORCH_API void PrepareDivisionForONNX(const std::shared_ptr<Graph>& graph);
```
Can you please move that to `torch/csrc/jit/passes/onnx`?

zou3519 committed:
Move prepare_division_for_onnx.cpp from passes/ to passes/onnx

zdevito commented on torch/onnx/utils.py:
```diff
@@ -141,6 +141,8 @@ def _optimize_graph(graph, operator_export_type):
     torch._C._jit_pass_lint(graph)
 
+    # onnx only supports tensors, but 1 / 2 = 0.5 and tensor(1) / tensor(2) = 0
+    torch._C._jit_pass_prepare_division_for_onnx(graph)
```
There is a RAW mode of ONNX export where everything gets turned into ATen-in-caffe2 operators (see the if statement below), but this pass will introduce onnx::Cast operators even in that mode. Is it not possible to do this fixup as part of `def div` in symbolic.py?

zou3519 replied:
Let me investigate and get back to you

zou3519 replied:
There isn't an easy way to do the fixup as part of the `def div` in symbolic.py -- EraseNumberTypes runs before `div` in symbolic gets called. Because of this, `div` doesn't have an easy way to check if it is dividing two numbers or two scalar tensors (it sees two scalar tensors).  The way to distinguish between the two is check the type of the outputs in the graph (LongTensor, LongTensor -> FloatTensor) but `div` doesn't get that information.

I ended up following @jamesr66a's advice and used `aten::_cast_Float` instead of `onnx::Cast` (thanks James!), let me know if this looks good.

jamesr66a commented on torch/csrc/jit/passes/onnx/prepare_division_for_onnx.cpp:
```diff
@@ -0,0 +1,36 @@
+        Value* longtensor = subgraph->insertNode(subgraph->createNumToTensor(input))->output();
+        // FLOAT = 1
+        // https://github.com/onnx/onnx/blob/6bedd27b0307c9295039bd847895a27275160a98/onnx/onnx.in.proto#L282
+        Node* cast = subgraph->create(onnx::Cast, {longtensor})->i_(attr::to, 1);
```
re: zach's comment below, make this emit an `aten::_cast_Float` operator. Then the ONNX pass later can correctly translate this based on which mode it's in (i.e. raw mode will pass that through, but ONNX mode will call the symbolic and insert what you have now)

jamesr66a reviewed (dismissed):
Fixup the ONNX pass and add TORCH_API to Node::matches (to make windows CI pass) then this should be good to go

zou3519 committed:
Address comments: use aten::_cast_Float instead of onnx::Cast

ezyang commented:
@pytorchbot retest this please

zou3519 commented:
@pytorchbot retest this please
